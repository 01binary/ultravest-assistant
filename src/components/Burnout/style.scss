@import '../../style/colors';

.longest {
	margin: 4px;
	padding: 2px 6px;
	background: $secondary-background-color;
	border-radius: 4px;
}
