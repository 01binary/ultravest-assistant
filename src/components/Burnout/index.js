import { h } from 'preact';
import presets from '../../config/burnoutPresets';
import getBurnoutPreset from '../../selectors/getBurnoutPreset';
import getBurnoutTime from '../../selectors/getBurnoutTime';
import Diagram from '../Diagram';
import Steps from '../Steps';
import style from './style';

/**
 * Display burnout preset calculated from flask size.
 * @param {object} flask - The flask props.
 * @param {bool} showSegments - Whether to show raw or formatted burnout steps.
 * @param {func} handleChangeView - Switches between raw and formatted steps.
 * @returns {JSX.Element} - A stateless component.
 */
const Burnout = ({
	flask,
	showSegments,
	handleChangeView
}) => {
	const preset = getBurnoutPreset(presets, flask);

	return (
		<article class={style.burnout}>
			<h2>burnout</h2>

			<output>
				<dl>
					<dt>preset</dt>
					<dd>
						{preset.diameter} X {preset.height} in
						{ getIsLastPreset(preset) &&
							<span class={style.longest}>longest</span>
						}
					</dd>

					<dt>time</dt>
					<dd>{getBurnoutTime(preset)} hours</dd>
				</dl>
			</output>

			<Diagram segments={preset.segments} />

			<Steps
				segments={preset.segments}
				showSegments={showSegments}
				handleChangeView={handleChangeView}
			/>
		</article>
	);
};

const getIsLastPreset = preset => (
	preset === presets[presets.length - 1] || null
);

export default Burnout;
