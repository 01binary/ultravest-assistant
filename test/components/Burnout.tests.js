import { h } from 'preact';
import { deep } from 'preact-render-spy';
import Burnout from '../../src/components/Burnout';
import withState from '../../src/composers';

describe('Burnout', () => {

	let wrapper;

	beforeAll(() => {
		wrapper = deep(h(withState(Burnout)));
	});

	afterAll(() => {
		wrapper = null;
	});

	test('should render', () => {
		expect(wrapper).toMatchSnapshot();
	});
});
