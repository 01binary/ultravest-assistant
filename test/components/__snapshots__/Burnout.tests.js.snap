// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Burnout with first burnout preset should not render last marker 1`] = `
preact-render-spy (1 nodes)
-------
<article class="burnout">
  <h2>burnout</h2>
  <output>
    <dl>
      <dt>preset</dt>
      <dd>3 X 3 in</dd>
      <dt>time</dt>
      <dd>8 hours</dd>
    </dl>
  </output>
  <section class="diagram">
    <a name="water-removal">
      <article class="segment raise">
        <figure class="temperature">
          <svg></svg>
          <figcaption>
            <h3>temp</h3>
            ambient
          </figcaption>
        </figure>
        <figure class="rate">
          <svg></svg>
          <figcaption>
            <h3>rate</h3>
            300°
            <span class="units">F/hr</span>
          </figcaption>
        </figure>
        <figure class="time">
          <svg></svg>
          <figcaption>
            <h3>hold</h3>
            1
            <span class="units">hour</span>
          </figcaption>
        </figure>
      </article>
    </a>
    <a name="thermal-transition">
      <article class="segment raise">
        <figure class="temperature">
          <svg></svg>
          <figcaption>
            <h3>temp</h3>
            700°
            <span class="units">F</span>
          </figcaption>
        </figure>
        <figure class="rate">
          <svg></svg>
          <figcaption>
            <h3>rate</h3>
            400°
            <span class="units">F/hr</span>
          </figcaption>
        </figure>
        <figure class="time">
          <svg></svg>
          <figcaption>
            <h3>hold</h3>
            1
            <span class="units">hour</span>
          </figcaption>
        </figure>
      </article>
    </a>
    <a name="pattern-removal">
      <article class="segment raise">
        <figure class="temperature">
          <svg></svg>
          <figcaption>
            <h3>temp</h3>
            1350°
            <span class="units">F</span>
          </figcaption>
        </figure>
        <figure class="rate">
          <svg></svg>
          <figcaption>
            <h3>rate</h3>
            325°
            <span class="units">F/hr</span>
          </figcaption>
        </figure>
        <figure class="time">
          <svg></svg>
          <figcaption>
            <h3>hold</h3>
            2
            <span class="units">hours</span>
          </figcaption>
        </figure>
      </article>
    </a>
    <a name="casting">
      <article class="segment lower">
        <figure class="temperature">
          <svg></svg>
          <figcaption>
            <h3>temp</h3>
            700°
            <span class="units">F</span>
          </figcaption>
        </figure>
        <figure class="rate">
          <svg></svg>
          <figcaption>
            <h3>rate</h3>
            700°
            <span class="units">F/hr</span>
          </figcaption>
        </figure>
        <figure class="time">
          <svg></svg>
          <figcaption>
            <h3>hold</h3>
            1
            <span class="units">hour</span>
          </figcaption>
        </figure>
      </article>
    </a>
  </section>
  <section class="steps">
    <input
      type="radio"
      name="program-view"
      id="steps"
      checked={true}
     />
    </input>
    <label for="steps">steps</label>
    <input
      type="radio"
      name="program-view"
      id="segments"
     />
    </input>
    <label for="segments">segments</label>
    <table>
      <thead>
        <tr>
          <th>segment</th>
          <th>rate</th>
          <th>temp</th>
          <th>hold</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <a href="#water-removal">water removal</a>
          </td>
          <td>300</td>
          <td>300</td>
          <td>1</td>
        </tr>
        <tr>
          <td>
            <a href="#thermal-transition">thermal transition</a>
          </td>
          <td>400</td>
          <td>700</td>
          <td>1</td>
        </tr>
        <tr>
          <td>
            <a href="#pattern-removal">pattern removal</a>
          </td>
          <td>325</td>
          <td>1350</td>
          <td>2</td>
        </tr>
        <tr>
          <td>
            <a href="#casting">casting</a>
          </td>
          <td>700</td>
          <td>700</td>
          <td>1</td>
        </tr>
      </tbody>
    </table>
    <ol>
      <li>
        <a href="#water-removal">
          raise to
          <span> 300</span>
          ° over 
          <span> 1 </span>
          hour
        </a>
      </li>
      <li>
        <a href="#water-removal">
          hold
          <span> 300</span>
          ° for 
          <span> 1 </span>
          hour
        </a>
      </li>
      <li>
        <a href="#thermal-transition">
          raise to
          <span> 700</span>
          ° over 
          <span> 1 </span>
          hour
        </a>
      </li>
      <li>
        <a href="#thermal-transition">
          hold
          <span> 700</span>
          ° for 
          <span> 1 </span>
          hour
        </a>
      </li>
      <li>
        <a href="#pattern-removal">
          raise to
          <span> 1350</span>
          ° over 
          <span> 2 </span>
          hours
        </a>
      </li>
      <li>
        <a href="#pattern-removal">
          hold
          <span> 1350</span>
          ° for 
          <span> 2 </span>
          hours
        </a>
      </li>
      <li>
        <a href="#casting">
          lower to
          <span> 700</span>
          ° over 
          <span> 0.9 </span>
          hour
        </a>
      </li>
      <li>
        <a href="#casting">
          hold
          <span> 700</span>
          ° for 
          <span> 1 </span>
          hour
        </a>
      </li>
    </ol>
  </section>
</article>

`;

exports[`Burnout with last burnout preset should render last marker 1`] = `
preact-render-spy (1 nodes)
-------
<article class="burnout">
  <h2>burnout</h2>
  <output>
    <dl>
      <dt>preset</dt>
      <dd>
        4 X 8 in
        <span class="longest">longest</span>
      </dd>
      <dt>time</dt>
      <dd>15 hours</dd>
    </dl>
  </output>
  <section class="diagram">
    <a name="water-removal">
      <article class="segment raise">
        <figure class="temperature">
          <svg></svg>
          <figcaption>
            <h3>temp</h3>
            ambient
          </figcaption>
        </figure>
        <figure class="rate">
          <svg></svg>
          <figcaption>
            <h3>rate</h3>
            300°
            <span class="units">F/hr</span>
          </figcaption>
        </figure>
        <figure class="time">
          <svg></svg>
          <figcaption>
            <h3>hold</h3>
            3
            <span class="units">hours</span>
          </figcaption>
        </figure>
      </article>
    </a>
    <a name="thermal-transition">
      <article class="segment raise">
        <figure class="temperature">
          <svg></svg>
          <figcaption>
            <h3>temp</h3>
            700°
            <span class="units">F</span>
          </figcaption>
        </figure>
        <figure class="rate">
          <svg></svg>
          <figcaption>
            <h3>rate</h3>
            200°
            <span class="units">F/hr</span>
          </figcaption>
        </figure>
        <figure class="time">
          <svg></svg>
          <figcaption>
            <h3>hold</h3>
            2
            <span class="units">hours</span>
          </figcaption>
        </figure>
      </article>
    </a>
    <a name="pattern-removal">
      <article class="segment raise">
        <figure class="temperature">
          <svg></svg>
          <figcaption>
            <h3>temp</h3>
            1350°
            <span class="units">F</span>
          </figcaption>
        </figure>
        <figure class="rate">
          <svg></svg>
          <figcaption>
            <h3>rate</h3>
            217°
            <span class="units">F/hr</span>
          </figcaption>
        </figure>
        <figure class="time">
          <svg></svg>
          <figcaption>
            <h3>hold</h3>
            3
            <span class="units">hours</span>
          </figcaption>
        </figure>
      </article>
    </a>
    <a name="casting">
      <article class="segment lower">
        <figure class="temperature">
          <svg></svg>
          <figcaption>
            <h3>temp</h3>
            700°
            <span class="units">F</span>
          </figcaption>
        </figure>
        <figure class="rate">
          <svg></svg>
          <figcaption>
            <h3>rate</h3>
            700°
            <span class="units">F/hr</span>
          </figcaption>
        </figure>
        <figure class="time">
          <svg></svg>
          <figcaption>
            <h3>hold</h3>
            2
            <span class="units">hours</span>
          </figcaption>
        </figure>
      </article>
    </a>
  </section>
  <section class="steps">
    <input
      type="radio"
      name="program-view"
      id="steps"
      checked={true}
     />
    </input>
    <label for="steps">steps</label>
    <input
      type="radio"
      name="program-view"
      id="segments"
     />
    </input>
    <label for="segments">segments</label>
    <table>
      <thead>
        <tr>
          <th>segment</th>
          <th>rate</th>
          <th>temp</th>
          <th>hold</th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td>
            <a href="#water-removal">water removal</a>
          </td>
          <td>300</td>
          <td>300</td>
          <td>3</td>
        </tr>
        <tr>
          <td>
            <a href="#thermal-transition">thermal transition</a>
          </td>
          <td>200</td>
          <td>700</td>
          <td>2</td>
        </tr>
        <tr>
          <td>
            <a href="#pattern-removal">pattern removal</a>
          </td>
          <td>217</td>
          <td>1350</td>
          <td>3</td>
        </tr>
        <tr>
          <td>
            <a href="#casting">casting</a>
          </td>
          <td>700</td>
          <td>700</td>
          <td>2</td>
        </tr>
      </tbody>
    </table>
    <ol>
      <li>
        <a href="#water-removal">
          raise to
          <span> 300</span>
          ° over 
          <span> 1 </span>
          hour
        </a>
      </li>
      <li>
        <a href="#water-removal">
          hold
          <span> 300</span>
          ° for 
          <span> 3 </span>
          hours
        </a>
      </li>
      <li>
        <a href="#thermal-transition">
          raise to
          <span> 700</span>
          ° over 
          <span> 2 </span>
          hours
        </a>
      </li>
      <li>
        <a href="#thermal-transition">
          hold
          <span> 700</span>
          ° for 
          <span> 2 </span>
          hours
        </a>
      </li>
      <li>
        <a href="#pattern-removal">
          raise to
          <span> 1350</span>
          ° over 
          <span> 3 </span>
          hours
        </a>
      </li>
      <li>
        <a href="#pattern-removal">
          hold
          <span> 1350</span>
          ° for 
          <span> 3 </span>
          hours
        </a>
      </li>
      <li>
        <a href="#casting">
          lower to
          <span> 700</span>
          ° over 
          <span> 0.9 </span>
          hour
        </a>
      </li>
      <li>
        <a href="#casting">
          hold
          <span> 700</span>
          ° for 
          <span> 2 </span>
          hours
        </a>
      </li>
    </ol>
  </section>
</article>

`;
